src/split_at.rs

```diff
@@ -72,113 +72,855 @@ pub unsafe trait SplitAt: KnownLayout<PointerMetadata = usize> {
+    /// portion must not overlap the right portion. For some dynamically sized
+    /// types, the padding that appears after the trailing slice field [is a
+    /// dynamic function of the trailing slice
+    /// length](KnownLayout#slice-dst-layout). Thus, for some types, this
+    /// condition is dependent on the length of the left portion.
+    #[must_use = "has no side effects"]
+    #[inline(always)]
+    pub unsafe fn via_unchecked(self) -> (&'a T, &'a [T::Elem]) {
+        // SAFETY: The aliasing of `self.into_ptr()` is not `Exclusive`, but the
+        // caller has promised that if `T` permits interior mutation then the
+        // left and right portions of `self` split at `l_len` do not overlap.
+        let (l, r) = unsafe { self.into_ptr().via_unchecked() };
+        (l.as_ref(), r.as_ref())
+    }
+
+    /// Reborrow `&self`, allowing the left and right parts to be borrowed
+    /// serially.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use zerocopy::{SplitAt, FromBytes, IntoBytes, network_endian::U16};
+    /// # use zerocopy_derive::*;
+    ///
+    /// #[derive(SplitAt, FromBytes, KnownLayout, Immutable, Debug)]
+    /// #[repr(C, align(2))]
+    /// struct Packet {
+    ///     length: U16,
+    ///     body: [u8],
+    /// }
+    ///
+    /// // These bytes encode a `Packet`.
+    /// let bytes = [
+    ///     3u16.to_be(),
+    ///     1u16.to_be(),
+    ///     2u16.to_be(),
+    ///     3u16.to_be(),
+    ///     4u16.to_be()
+    /// ];
+    ///
+    /// let packet = Packet::ref_from_bytes(bytes.as_bytes()).unwrap();
+    ///
+    /// assert_eq!(packet.length, 3);
+    /// assert_eq!(packet.body, [0, 1, 0, 2, 0, 3, 0, 4]);
+    ///
+    /// // Attempt to split `packet` at `length`.
+    /// let mut split = packet.split_at(packet.length.into()).unwrap();
+    ///
+    /// // Borrow the left part of the split.
+    /// {
+    ///     let split = split.reborrow();
+    ///     let packet = split.left();
+    ///
+    ///     assert_eq!(packet.length, 3);
+    ///     assert_eq!(packet.body, [0, 1, 0]);
+    /// }
+    ///
+    /// // Borrow the right part of the split.
+    /// {
+    ///     let split = split.reborrow();
+    ///     let rest = split.right();
+    ///
+    ///     assert_eq!(rest, [2, 0, 3, 0, 4]);
+    /// }
+    /// ```
+    ///
+    /// This method cannot be used to concurrently borrow the left and right
+    /// parts; e.g.:
+    ///
+    /// ```compile_fail,E0499
+    /// use zerocopy::SplitAt;
+    ///
+    /// let mut split = SplitAt::split_at(&[0, 1, 2, 3][..], 2).unwrap();
+    ///
+    /// let left = split.reborrow().left();
+    /// let right = split.reborrow().right();
+    ///
+    /// println!("{:?}", left);
+    /// println!("{:?}", right);
+    /// ```
+    #[must_use = "has no side effects"]
+    #[inline(always)]
+    pub fn reborrow<'b>(&'b mut self) -> Split<&'b T> {
+        // SAFETY: `&*self.source` points to exactly `self.source` and thus
```

Comment:
It wigs me out that this safety proof doesn't mention lifetimes at all, despite the fact that we explicitly advertise that you can't reborrow twice overlapping (and that that's a soundness requirement). Maybe we need to revisit the safety invariant on `Split` or the safety precondition on `Split::new`?

Same with the other `fn reborrow`.

Reply:
Yeah, I agree something is fishy here. Similarly, it would be nice if `SplitAt`'s invariant obviously precluded implementing `Copy`/`Clone`. 

It's not quite about lifetimes, I don't think, but rather about shadowing; e.g., on `SplitAt::reborrow`:
```rust
        // SAFETY: `&mut *self.source` points to exactly `self.source` and thus
        // maintains the invariants of `self` with respect to `l_len`. The
        // reborrowed `&mut *self.source` shadows `&mut self`'s source, ensuring
        // that it is the sole active reference from which the left and right
        // parts may be accessed.
```

...but I'm not immediately sure what aspects of this need to be uplifted into `SplitAt`.

Reply:
Discussed offline, and decided that `left`, `right`, and `reborrow` are unsound for both `&` and `&mut`. In particular:
- In the `&` case, `left` can overlap with pre-existing `&` references to `T`, and so if there's padding and `T: !Immutable`, then this exposes unsoundness
- In both the `&` case (where `T: !Immutable`) and the `&mut` case, if you use `left` to overwrite padding, then once you have dropped the reference, you un-shadow the original, which now spuriously points to uninit bytes where there should be none

Thus, we're going to remove `left`, `right`, and `reborrow`.